Thunderbird API

MelissaAutumn opened the pull request:
Woah we have an addons linter for Thunderbird? Wild!

@jobisoft 

A few minor tweaks were needed to get this working with the Thunderbird API. I've used the latest tip for the schema, let me know if there's a specific revision you'd like. (I think we want tip for daily compatability on ATN though.) 

I have a feeling my error message change will break localized content, but I'm not sure we display localized error/warnings on ATN. Can I get a confirmation on that?

I've tested it with the sample extensions, and I think it's working as intended, but I'd like a second pair of eyes on it. You'll just need to `npm install` and `npm run build` and it should be in `./bin/addons-linter`. 

You can ignore anything in the /imported/ folder as thats...imported via scripts.

MelissaAutumn committed:
Added a script to split tb schemas by namespace, and minor adjustments to addons-linter to accommodate that.

MelissaAutumn committed:
Rename and re-work the split_namespaces script to download apis

MelissaAutumn committed:
Fix up some code so it doesn't break with Thunderbird's API

MelissaAutumn committed:
Add Thunderbird Schema

MelissaAutumn committed:
Add a build option to ignore privileged errors

MelissaAutumn committed:
Merge branch 'master' into thunderbird

MelissaAutumn committed:
Readme instructions

jobisoft commented on src/schema/firefox-schemas-import.js:
```diff
@@ -552,12 +555,19 @@ inner.normalizeSchema = (schemas, file) => {
       // we want to expand those definitions otherwise they will be converted into
       // a $merge form, which is only used in the manifest validation and so the
       // eslint rules that look for the API method definitions will not work as expected.
+      const obj = apiNamespaceSchemas.find((schema) => schema.namespace === $import);
+
+      if (obj === undefined) {
+        return {};
+      }
+
```
We need to be careful here. action API is MV3 and browser_action is MV2 and there will be differences between them.

Is the linter checking for MV2 and MV3?

MelissaAutumn replied:
Both MV2 and MV3. It splits it properly, I'll make a comment when I push the newer fixed version.

jobisoft commented on src/messages/javascript.js:
```diff
@@ -152,7 +154,7 @@ function temporaryAPI(api) {
     description: i18n._(oneLine`This API can cause issues when loaded
-      temporarily using about:debugging in Firefox unless you specify
+      temporarily using about:debugging in Thunderbird unless you specify
       applications|browser_specific_settings > gecko > id in the manifest.
```
`applications` is deprecated and we should use `browser_specific_settings` only. Maybe remove it here, so we do not have to deal with it later anymore?

jobisoft commented on src/messages/javascript.js:
```diff
@@ -152,7 +154,7 @@ function temporaryAPI(api) {
     message: i18n._(`"${api}" can cause issues when loaded temporarily`),
     description: i18n._(oneLine`This API can cause issues when loaded
-      temporarily using about:debugging in Firefox unless you specify
+      temporarily using about:debugging in Thunderbird unless you specify
```
`applications` is deprecated and we should use `browser_specific_settings` only. Maybe remove it here, so we do not have to deal with it later anymore?

jobisoft commented:
I did not know that we had this. Nice! I think we need to put some brain behind picking the right schema, or not?

For my understanding: This is a stand-alone tool for developers and will not be used by ATN to check submissions, right?

When linting an add-on, it includes strict_min/max_version and that should tell us, what schema version should be used. I even think, that we should check against multiple definitions. For example, an add-on which has only `strict_min_version: 78.0` should be checked against ESR78, ESR91, ESR102, BETA and CENTRAL. Because that is what the add-on claims to support. So the linter should include these schema files.

But developers may include conditional function calls, to deal with version dependent differences. On the other hand, we try to keep those low, because WebExtensions aim to be backward compatible. Except MV3, which we will use for mayor API deprecations and changes.

So, I do not really know what to do here. What do you think?

The idea here is that the user of that tool does not have to download the sources himself, but all schema files are shipped with it, right?

jobisoft commented:
My github does not like `\addons-linter\tests\fixtures\coinminers`, something we need to worry about?

jobisoft commented:
I get the following error:

```
bin/addons-linter /mnt/c/Users/John/Downloads/mailhops-4.4.0-tb.xpi
can't resolve reference search#/definitions/OptionalPermissionNoPrompt from id manifest
(node:388) UnhandledPromiseRejectionWarning: Error: can't resolve reference search#/definitions/OptionalPermissionNoPrompt from id manifest
    at Object.code (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/vocabularies/core/ref.ts:19:39)
    at keywordCode (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/validate/index.ts:532:9)
    at /mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/validate/index.ts:228:21
    at CodeGen.code (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/codegen/index.ts:525:33)
    at CodeGen.block (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/codegen/index.ts:680:20)
    at schemaKeywords (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/validate/index.ts:228:9)
    at typeAndKeywords (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/validate/index.ts:161:3)
    at subSchemaObjCode (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/validate/index.ts:147:3)
    at subschemaCode (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/validate/index.ts:124:7)
    at KeywordCxt.subschema (/mnt/c/Users/John/Documents/GitHub/addons-linter/node_modules/ajv/lib/compile/validate/index.ts:500:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:388) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:388) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

MelissaAutumn committed:
Create a schema/updates json for each namespace we need to split off, and remove my custom splitting solution.

We can now remove the action/browserAction work-around, and I've also fixed the missing commands.json/pkcs10.json.

MelissaAutumn commented:
I think you'll need at least node 18 to run this

jobisoft commented on src/messages/javascript.js:
```diff
@@ -152,7 +154,7 @@ function temporaryAPI(api) {
     message: i18n._(`"${api}" can cause issues when loaded temporarily`),
     description: i18n._(oneLine`This API can cause issues when loaded
-      temporarily using about:debugging in Firefox unless you specify
+      temporarily using about:debugging in Thunderbird unless you specify
       applications|browser_specific_settings > gecko > id in the manifest.
```
I know this is not part of your changeset, but "since we are here": 

`applications` is deprecated and we should use `browser_specific_settings` only. Maybe remove it here, so we do not have to deal with it later anymore?

MelissaAutumn replied:
Sounds good, I'm just double checking to see if I can non-destructively modify the strings first.

MelissaAutumn committed:
More Firefox -> Thunderbird strings. Also add a future proof script for when i18n support works (or when we can get it working...)

MelissaAutumn committed:
Update readme

MelissaAutumn commented:
Well I fell down a rabbit hole trying to get i18n / gettext working. It doesn't seem to be functional, so I've went ahead and modified the strings directly. I've added a patch script which will work whenever i18n/gettext is working again.

The imported API also currently matches `THUNDERBIRD_113_0b1_RELEASE`. (As the Firefox addons-linter matches their beta releases.)

jobisoft commented on src/schema/firefox-schemas-import.js:
```diff
@@ -38,6 +38,8 @@ const schemaRegexes = [
   new RegExp('browser/components/extensions/schemas/.*\\.json'),
```
We do not load these and they should not be used for linting (Thunderbird really does not know them).

jobisoft replied:
Only later I saw we do clone `/browser/components/extensions/schemas/pkcs11.json`. If this is the reason why the browser folder is listed here, could you add a comment to explain that?

MelissaAutumn replied:
Sounds good!

jobisoft commented on tb-scripts/download_apis.py:
```diff
@@ -0,0 +1,119 @@
+    files = {
+        mozilla_central_folder: [
+            'toolkit/',
+            'browser/components/extensions/schemas/commands.json',
```
We have our own copy of `commands.json` and it should not be merged with the firefox version.

However, according to this file:
https://searchfox.org/comm-central/source/mail/components/extensions/jar.mn#58

we do clone `pkcs11.json` which is unusual, but I guess the rationale is to not have to monitor upstream for changes. But having that hardcoded here may get forgotten later. I will ask rob, if we can copy that file to c-c during built-time. We will have to keep including it here for now.

jobisoft commented on README.md:
```diff
@@ -8,6 +8,39 @@ It can also be used as a standalone binary and library.
 
 You can find more information about the linter and it's implemented rules in our [documentation](https://mozilla.github.io/addons-linter/).
 
```
The `addons-linter` section does not reflect the disconnect from Firefox. I think it can only be used as a standalone binary and library, correct? Is it used by ATN?

jobisoft replied:
The CircleCI is still pointing to the firefox repo, correct? If it does not work for us, I guess we can just remove it for now.

jobisoft replied:
There are more places in this readme, which need an update to reflect this is a different repo and not the firefox one. Since these places are not changes, the review tool does not allow me to add remarks there, or at least I am not able to add them. :-( 

For example the `Quick Start` section has the firefox repo link hardcoded.

MelissaAutumn replied:
I'll adjust the readme. This will also be used on ATN.

MelissaAutumn commented:
Hey @jobisoft I'm putting this PR on hold until I start working on the addons server stuff hopefully in the fall. Ran into some functionality changes that were needed as well, and too many things and too little time to get this out the door right now. 😅 

I will perform the changes you wanted for ATN on ATN.
